File: htui/src/main/java/com/ht117/htui/loading/CircleLoadingView.java
package com.ht117.htui.loading;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.os.Handler;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;

import com.ht117.htui.R;
import com.ht117.htui.Utils;

/**
 * Created by steve on 7/9/17.
 */

public class CircleLoadingView extends View {

    private int coLeft;
    private int coRight;
    private int coTop;
    private int coBot;
    private int interval;
    private int strokeWidth;
    private int startDeg;
    private final int SWEEP = 90;

    private Paint painterLeft;
    private Paint painterTop;
    private Paint painterRight;
    private Paint painterBot;

    private Handler handler;

    public CircleLoadingView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        initView(attrs);
        handler = new Handler();
    }

    private void initView(AttributeSet attrs) {
        TypedArray array = null;
        try {
            array = getContext().obtainStyledAttributes(attrs, R.styleable.CircleLoadingView);

            coLeft = array.getColor(R.styleable.CircleLoadingView_clv_colorLeft, Color.BLUE);
            coTop = array.getColor(R.styleable.CircleLoadingView_clv_colorTop, Color.YELLOW);
            coRight = array.getColor(R.styleable.CircleLoadingView_clv_colorRight, Color.GREEN);
            coBot = array.getColor(R.styleable.CircleLoadingView_clv_colorBot, Color.RED);
            interval = array.getInt(R.styleable.CircleLoadingView_clv_interval, Utils.SECOND / 10);
            strokeWidth = array.getInt(R.styleable.CircleLoadingView_clv_strokeWidth, Utils.DEF_STROKE_WIDTH);
            startDeg = Utils.ZERO;

            painterLeft = new Paint(Paint.ANTI_ALIAS_FLAG);
            painterLeft.setStyle(Paint.Style.STROKE);
            painterLeft.setStrokeWidth(strokeWidth);
            painterLeft.setColor(coLeft);

            painterRight = new Paint(Paint.ANTI_ALIAS_FLAG);
            painterRight.setStyle(Paint.Style.STROKE);
            painterRight.setStrokeWidth(strokeWidth);
            painterRight.setColor(coRight);

            painterTop = new Paint(Paint.ANTI_ALIAS_FLAG);
            painterTop.setStyle(Paint.Style.STROKE);
            painterTop.setStrokeWidth(strokeWidth);
            painterTop.setColor(coTop);

            painterBot = new Paint(Paint.ANTI_ALIAS_FLAG);
            painterBot.setStyle(Paint.Style.STROKE);
            painterBot.setStrokeWidth(strokeWidth);
            painterBot.setColor(coBot);

        } finally {
            if (array != null) {
                array.recycle();
            }
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        int left = getPaddingLeft();
        int top = getPaddingTop();
        int width = getWidth();
        int height = getHeight();

        int padding = Utils.PADDING;
        RectF rect = new RectF(left + padding, top + padding, left + width - padding, top + height - padding);

        canvas.drawArc(rect, startDeg, SWEEP, false, painterLeft);
        canvas.drawArc(rect, startDeg + SWEEP, SWEEP, false, painterTop);
        canvas.drawArc(rect, startDeg + SWEEP * 2, SWEEP, false, painterRight);
        canvas.drawArc(rect, startDeg + SWEEP * 3, SWEEP, false, painterBot);
        handler.postDelayed(handleCtrl, interval);
    }

    private final Runnable handleCtrl = new Runnable() {
        @Override
        public void run() {
            int temp = coLeft;
            coLeft = coTop;
            coTop = coRight;
            coRight = coBot;
            coBot = temp;
            startDeg = (startDeg + 1) % Utils.TWO_PI;

            painterLeft.setColor(coLeft);
            painterTop.setColor(coTop);
            painterRight.setColor(coRight);
            painterBot.setColor(coBot);
            invalidate();
        }
    };
}

File: htui/src/main/java/com/ht117/htui/loading/CircleLoadingView.kt
package com.ht117.htui.loading

import android.content.Context
import android.content.res.TypedArray
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.util.AttributeSet
import android.view.View
import androidx.annotation.Nullable
import com.ht117.htui.R
import com.ht117.htui.Utils

/**
 * Created by steve on 7/9/17.
 */

class CircleLoadingView(context: Context, @Nullable attrs: AttributeSet) : View(context, attrs) {

    private var coLeft: Int = 0
    private var coRight: Int = 0
    private var coTop: Int = 0
    private var coBot: Int = 0
    private var interval: Int = 0
    private var strokeWidth: Int = 0
    private var startDeg: Int = 0
    private val SWEEP = 90

    private var painterLeft: Paint? = null
    private var painterTop: Paint? = null
    private var painterRight: Paint? = null
    private var painterBot: Paint? = null

    private val handleCtrl = Runnable {
        val temp = coLeft
        coLeft = coTop
        coTop = coRight
        coRight = coBot
        coBot = temp
        startDeg = (startDeg + 1) % Utils.TWO_PI

        painterLeft!!.color = coLeft
        painterTop!!.color = coTop
        painterRight!!.color = coRight
        painterBot!!.color = coBot
        invalidate()
    }

    init {
        initView(attrs)
    }

    private fun initView(attrs: AttributeSet) {
        var array: TypedArray? = null
        try {
            array = context.obtainStyledAttributes(attrs, R.styleable.CircleLoadingView)

            coLeft = array!!.getColor(R.styleable.CircleLoadingView_clv_colorLeft, Color.BLUE)
            coTop = array.getColor(R.styleable.CircleLoadingView_clv_colorTop, Color.YELLOW)
            coRight = array.getColor(R.styleable.CircleLoadingView_clv_colorRight, Color.GREEN)
            coBot = array.getColor(R.styleable.CircleLoadingView_clv_colorBot, Color.RED)
            interval = array.getInt(R.styleable.CircleLoadingView_clv_interval, Utils.SECOND / 10)
            strokeWidth = array.getInt(R.styleable.CircleLoadingView_clv_strokeWidth, Utils.DEF_STROKE_WIDTH)
            startDeg = Utils.ZERO
        } finally {
            array?.recycle()
        }

        painterLeft = Paint(Paint.ANTI_ALIAS_FLAG)
        painterLeft!!.style = Paint.Style.STROKE
        painterLeft!!.strokeWidth = strokeWidth.toFloat()
        painterLeft!!.color = coLeft

        painterRight = Paint(Paint.ANTI_ALIAS_FLAG)
        painterRight!!.style = Paint.Style.STROKE
        painterRight!!.strokeWidth = strokeWidth.toFloat()
        painterRight!!.color = coRight

        painterTop = Paint(Paint.ANTI_ALIAS_FLAG)
        painterTop!!.style = Paint.Style.STROKE
        painterTop!!.strokeWidth = strokeWidth.toFloat()
        painterTop!!.color = coTop

        painterBot = Paint(Paint.ANTI_ALIAS_FLAG)
        painterBot!!.style = Paint.Style.STROKE
        painterBot!!.strokeWidth = strokeWidth.toFloat()
        painterBot!!.color = coBot

    }

    override fun onDraw(canvas: Canvas) {
        val left = paddingLeft
        val top = paddingTop
        val width = width
        val height = height

        val padding = Utils.PADDING
        val rect = RectF((left + padding).toFloat(), (top + padding).toFloat(), (left + width - padding).toFloat(), (top + height - padding).toFloat())

        canvas.drawArc(rect, startDeg.toFloat(), SWEEP.toFloat(), false, painterLeft!!)
        canvas.drawArc(rect, (startDeg + SWEEP).toFloat(), SWEEP.toFloat(), false, painterTop!!)
        canvas.drawArc(rect, (startDeg + SWEEP * 2).toFloat(), SWEEP.toFloat(), false, painterRight!!)
        canvas.drawArc(rect, (startDeg + SWEEP * 3).toFloat(), SWEEP.toFloat(), false, painterBot!!)
        handler.postDelayed(handleCtrl, interval.toLong())
    }
}
